File: composeApp/src/androidMain/java/org/application/shikiapp/shared/SpannedToHtmlConverter.java
package org.application.shikiapp.shared;

import static com.google.common.base.Preconditions.checkNotNull;

import android.graphics.Typeface;
import android.text.Html;
import android.text.Spanned;
import android.text.style.AbsoluteSizeSpan;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.RelativeSizeSpan;
import android.text.style.StrikethroughSpan;
import android.text.style.StyleSpan;
import android.text.style.TypefaceSpan;
import android.text.style.UnderlineSpan;
import android.util.SparseArray;

import androidx.annotation.Nullable;
import androidx.annotation.OptIn;
import androidx.media3.common.text.HorizontalTextInVerticalContextSpan;
import androidx.media3.common.text.RubySpan;
import androidx.media3.common.text.TextAnnotation;
import androidx.media3.common.text.TextEmphasisSpan;
import androidx.media3.common.util.UnstableApi;
import androidx.media3.common.util.Util;

import com.google.common.collect.ImmutableMap;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

final class SpannedToHtmlConverter {

  private static final Pattern NEWLINE_PATTERN = Pattern.compile("(&#13;)?&#10;");

  private SpannedToHtmlConverter() {}

  @OptIn(markerClass = UnstableApi.class)
  public static HtmlAndCss convert(@Nullable CharSequence text, float displayDensity) {
    if (text == null) {
      return new HtmlAndCss("", ImmutableMap.of());
    }
    if (!(text instanceof Spanned spanned)) {
      return new HtmlAndCss(escapeHtml(text), ImmutableMap.of());
    }

      Set<Integer> backgroundColors = new HashSet<>();
    for (BackgroundColorSpan backgroundColorSpan :
            spanned.getSpans(0, spanned.length(), BackgroundColorSpan.class)) {
      backgroundColors.add(backgroundColorSpan.getBackgroundColor());
    }
    HashMap<String, String> cssRuleSets = new HashMap<>();
    for (int backgroundColor : backgroundColors) {
      cssRuleSets.put(
              HtmlUtils.cssAllClassDescendantsSelector("bg_" + backgroundColor),
              Util.formatInvariant("background-color:%s;", HtmlUtils.toCssRgba(backgroundColor)));
    }

    SparseArray<Transition> spanTransitions = findSpanTransitions(spanned, displayDensity);
    StringBuilder html = new StringBuilder(spanned.length());
    int previousTransition = 0;
    for (int i = 0; i < spanTransitions.size(); i++) {
      int index = spanTransitions.keyAt(i);
      html.append(escapeHtml(spanned.subSequence(previousTransition, index)));

      Transition transition = spanTransitions.get(index);
      transition.spansRemoved.sort(SpanInfo.FOR_CLOSING_TAGS);
      for (SpanInfo spanInfo : transition.spansRemoved) {
        html.append(spanInfo.closingTag);
      }
      transition.spansAdded.sort(SpanInfo.FOR_OPENING_TAGS);
      for (SpanInfo spanInfo : transition.spansAdded) {
        html.append(spanInfo.openingTag);
      }
      previousTransition = index;
    }

    html.append(escapeHtml(spanned.subSequence(previousTransition, spanned.length())));

    return new HtmlAndCss(html.toString(), cssRuleSets);
  }

  private static SparseArray<Transition> findSpanTransitions(
          Spanned spanned, float displayDensity) {
    SparseArray<Transition> spanTransitions = new SparseArray<>();

    for (Object span : spanned.getSpans(0, spanned.length(), Object.class)) {
      @Nullable String openingTag = getOpeningTag(span, displayDensity);
      @Nullable String closingTag = getClosingTag(span);
      int spanStart = spanned.getSpanStart(span);
      int spanEnd = spanned.getSpanEnd(span);
      if (openingTag != null) {
        checkNotNull(closingTag);
        SpanInfo spanInfo = new SpanInfo(spanStart, spanEnd, openingTag, closingTag);
        getOrCreate(spanTransitions, spanStart).spansAdded.add(spanInfo);
        getOrCreate(spanTransitions, spanEnd).spansRemoved.add(spanInfo);
      }
    }

    return spanTransitions;
  }

  @OptIn(markerClass = UnstableApi.class)
  @Nullable
  private static String getOpeningTag(Object span, float displayDensity) {
      switch (span) {
          case StrikethroughSpan ignored -> {
              return "<span style='text-decoration:line-through;'>";
          }
          case ForegroundColorSpan colorSpan -> {
              return Util.formatInvariant(
                      "<span style='color:%s;'>", HtmlUtils.toCssRgba(colorSpan.getForegroundColor()));
          }
          case BackgroundColorSpan colorSpan -> {
              return Util.formatInvariant("<span class='bg_%s'>", colorSpan.getBackgroundColor());
          }
          case HorizontalTextInVerticalContextSpan ignored -> {
              return "<span style='text-combine-upright:all;'>";
          }
          case AbsoluteSizeSpan absoluteSizeSpan -> {
              float sizeCssPx =
                      absoluteSizeSpan.getDip()
                              ? absoluteSizeSpan.getSize()
                              : absoluteSizeSpan.getSize() / displayDensity;
              return Util.formatInvariant("<span style='font-size:%.2fpx;'>", sizeCssPx);
          }
          case RelativeSizeSpan relativeSizeSpan -> {
              return Util.formatInvariant(
                      "<span style='font-size:%.2f%%;'>", relativeSizeSpan.getSizeChange() * 100);
          }
          case TypefaceSpan typefaceSpan -> {
              @Nullable String fontFamily = typefaceSpan.getFamily();
              return fontFamily != null
                      ? Util.formatInvariant("<span style='font-family:\"%s\";'>", fontFamily)
                      : null;
          }
          case StyleSpan styleSpan -> {
              return switch (styleSpan.getStyle()) {
                  case Typeface.BOLD -> "<b>";
                  case Typeface.ITALIC -> "<i>";
                  case Typeface.BOLD_ITALIC -> "<b><i>";
                  default -> null;
              };
          }
          case RubySpan rubySpan -> {
              return switch (rubySpan.position) {
                  case TextAnnotation.POSITION_BEFORE -> "<ruby style='ruby-position:over;'>";
                  case TextAnnotation.POSITION_AFTER -> "<ruby style='ruby-position:under;'>";
                  case TextAnnotation.POSITION_UNKNOWN -> "<ruby style='ruby-position:unset;'>";
                  default -> null;
              };
          }
          case UnderlineSpan ignored -> {
              return "<u>";
          }
          case TextEmphasisSpan textEmphasisSpan -> {
              String style = getTextEmphasisStyle(textEmphasisSpan.markShape, textEmphasisSpan.markFill);
              String position = getTextEmphasisPosition(textEmphasisSpan.position);
              return Util.formatInvariant(
                      "<span style='-webkit-text-emphasis-style:%1$s;text-emphasis-style:%1$s;"
                              + "-webkit-text-emphasis-position:%2$s;text-emphasis-position:%2$s;"
                              + "display:inline-block;'>",
                      style, position);
          }
          case null, default -> {
              return null;
          }
      }
  }

  @OptIn(markerClass = UnstableApi.class)
  @Nullable
  private static String getClosingTag(Object span) {
    if (span instanceof StrikethroughSpan
            || span instanceof ForegroundColorSpan
            || span instanceof BackgroundColorSpan
            || span instanceof HorizontalTextInVerticalContextSpan
            || span instanceof AbsoluteSizeSpan
            || span instanceof RelativeSizeSpan
            || span instanceof TextEmphasisSpan) {
      return "</span>";
    } else if (span instanceof TypefaceSpan) {
      @Nullable String fontFamily = ((TypefaceSpan) span).getFamily();
      return fontFamily != null ? "</span>" : null;
    } else if (span instanceof StyleSpan) {
      switch (((StyleSpan) span).getStyle()) {
        case Typeface.BOLD:
          return "</b>";
        case Typeface.ITALIC:
          return "</i>";
        case Typeface.BOLD_ITALIC:
          return "</i></b>";
      }
    } else if (span instanceof RubySpan rubySpan) {
        return "<rt>" + escapeHtml(rubySpan.rubyText) + "</rt></ruby>";
    } else if (span instanceof UnderlineSpan) {
      return "</u>";
    }
    return null;
  }

  @OptIn(markerClass = UnstableApi.class)
  private static String getTextEmphasisStyle(
          @TextEmphasisSpan.MarkShape int shape, @TextEmphasisSpan.MarkFill int fill) {
    StringBuilder builder = new StringBuilder();
    switch (fill) {
      case TextEmphasisSpan.MARK_FILL_FILLED:
        builder.append("filled ");
        break;
      case TextEmphasisSpan.MARK_FILL_OPEN:
        builder.append("open ");
        break;
      case TextEmphasisSpan.MARK_FILL_UNKNOWN:
      default:
        break;
    }

    switch (shape) {
      case TextEmphasisSpan.MARK_SHAPE_CIRCLE:
        builder.append("circle");
        break;
      case TextEmphasisSpan.MARK_SHAPE_DOT:
        builder.append("dot");
        break;
      case TextEmphasisSpan.MARK_SHAPE_SESAME:
        builder.append("sesame");
        break;
      case TextEmphasisSpan.MARK_SHAPE_NONE:
        builder.append("none");
        break;
      default:
        builder.append("unset");
        break;
    }
    return builder.toString();
  }

  @OptIn(markerClass = UnstableApi.class)
  private static String getTextEmphasisPosition(@TextAnnotation.Position int position) {
      if (position == TextAnnotation.POSITION_AFTER) {
          return "under left";
      }
      return "over right";
  }

  private static Transition getOrCreate(SparseArray<Transition> transitions, int key) {
    @Nullable Transition transition = transitions.get(key);
    if (transition == null) {
      transition = new Transition();
      transitions.put(key, transition);
    }
    return transition;
  }

  private static String escapeHtml(CharSequence text) {
    String escaped = Html.escapeHtml(text);
    return NEWLINE_PATTERN.matcher(escaped).replaceAll("<br>");
  }

  public static class HtmlAndCss {

    public final String html;

    public final Map<String, String> cssRuleSets;

    private HtmlAndCss(String html, Map<String, String> cssRuleSets) {
      this.html = html;
      this.cssRuleSets = cssRuleSets;
    }
  }

  private record SpanInfo(int start, int end, String openingTag, String closingTag) {
      private static final Comparator<SpanInfo> FOR_OPENING_TAGS =
              (info1, info2) -> {
                int result = Integer.compare(info2.end, info1.end);
                if (result != 0) {
                  return result;
                }
                result = info1.openingTag.compareTo(info2.openingTag);
                if (result != 0) {
                  return result;
                }
                return info1.closingTag.compareTo(info2.closingTag);
              };

      private static final Comparator<SpanInfo> FOR_CLOSING_TAGS =
              (info1, info2) -> {
                int result = Integer.compare(info2.start, info1.start);
                if (result != 0) {
                  return result;
                }
                result = info2.openingTag.compareTo(info1.openingTag);
                if (result != 0) {
                  return result;
                }
                return info2.closingTag.compareTo(info1.closingTag);
              };

  }

  private static final class Transition {
    private final List<SpanInfo> spansAdded;
    private final List<SpanInfo> spansRemoved;

    public Transition() {
      this.spansAdded = new ArrayList<>();
      this.spansRemoved = new ArrayList<>();
    }
  }
}

File: composeApp/src/androidMain/kotlin/org/application/shikiapp/shared/utils/ui/subtitles/SpannedToHtmlConverter.kt
package org.application.shikiapp.shared.utils.ui.subtitles

import android.graphics.Typeface
import android.text.Html
import android.text.Spanned
import android.text.style.AbsoluteSizeSpan
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.text.style.RelativeSizeSpan
import android.text.style.StrikethroughSpan
import android.text.style.StyleSpan
import android.text.style.TypefaceSpan
import android.text.style.UnderlineSpan
import android.util.SparseArray
import androidx.annotation.OptIn
import androidx.core.util.size
import androidx.media3.common.text.HorizontalTextInVerticalContextSpan
import androidx.media3.common.text.RubySpan
import androidx.media3.common.text.TextAnnotation
import androidx.media3.common.text.TextEmphasisSpan
import androidx.media3.common.util.UnstableApi
import androidx.media3.common.util.Util

object SpannedToHtmlConverter {

    private val NEWLINE_REGEX = Regex("(&#13;)?&#10;")

    @OptIn(UnstableApi::class)
    fun convert(text: CharSequence?, displayDensity: Float): HtmlAndCss {
        if (text == null) {
            return HtmlAndCss("", emptyMap())
        }
        if (text !is Spanned) {
            return HtmlAndCss(escapeHtml(text), emptyMap())
        }

        val backgroundColors = text.getSpans(0, text.length, BackgroundColorSpan::class.java)
            .map { it.backgroundColor }
            .toSet()

        val cssRuleSets = backgroundColors.associate { backgroundColor ->
            HtmlUtils.cssAllClassDescendantsSelector("bg_$backgroundColor") to
                    Util.formatInvariant(
                        "background-color:%s;",
                        HtmlUtils.toCssRgba(backgroundColor)
                    )
        }

        val spanTransitions = findSpanTransitions(text, displayDensity)
        val html = StringBuilder(text.length)
        var previousTransition = 0

        for (i in 0 until spanTransitions.size) {
            val index = spanTransitions.keyAt(i)
            html.append(escapeHtml(text.subSequence(previousTransition, index)))

            val transition = spanTransitions.valueAt(i)
            transition.spansRemoved.sortWith(SpanInfo.FOR_CLOSING_TAGS)
            for (spanInfo in transition.spansRemoved) {
                html.append(spanInfo.closingTag)
            }

            transition.spansAdded.sortWith(SpanInfo.FOR_OPENING_TAGS)
            for (spanInfo in transition.spansAdded) {
                html.append(spanInfo.openingTag)
            }
            previousTransition = index
        }

        html.append(escapeHtml(text.subSequence(previousTransition, text.length)))

        return HtmlAndCss(html.toString(), cssRuleSets)
    }

    private fun findSpanTransitions(
        spanned: Spanned,
        displayDensity: Float
    ): SparseArray<Transition> {
        val spanTransitions = SparseArray<Transition>()

        for (span in spanned.getSpans(0, spanned.length, Any::class.java)) {
            val openingTag = getOpeningTag(span, displayDensity)
            val closingTag = getClosingTag(span)
            val spanStart = spanned.getSpanStart(span)
            val spanEnd = spanned.getSpanEnd(span)

            if (openingTag != null) {
                requireNotNull(closingTag)
                val spanInfo = SpanInfo(spanStart, spanEnd, openingTag, closingTag)
                spanTransitions.getOrCreate(spanStart).spansAdded.add(spanInfo)
                spanTransitions.getOrCreate(spanEnd).spansRemoved.add(spanInfo)
            }
        }

        return spanTransitions
    }

    @OptIn(UnstableApi::class)
    private fun getOpeningTag(span: Any, displayDensity: Float): String? {
        return when (span) {
            is StrikethroughSpan -> "<span style='text-decoration:line-through;'>"
            is ForegroundColorSpan -> Util.formatInvariant(
                "<span style='color:%s;'>",
                HtmlUtils.toCssRgba(span.foregroundColor)
            )

            is BackgroundColorSpan -> Util.formatInvariant(
                "<span class='bg_%s'>",
                span.backgroundColor
            )

            is HorizontalTextInVerticalContextSpan -> "<span style='text-combine-upright:all;'>"
            is AbsoluteSizeSpan -> {
                val sizeCssPx = if (span.dip) span.size.toFloat() else span.size / displayDensity
                Util.formatInvariant("<span style='font-size:%.2fpx;'>", sizeCssPx)
            }

            is RelativeSizeSpan -> Util.formatInvariant(
                "<span style='font-size:%.2f%%;'>",
                span.sizeChange * 100
            )

            is TypefaceSpan -> span.family?.let {
                Util.formatInvariant(
                    "<span style='font-family:\"%s\";'>",
                    it
                )
            }

            is StyleSpan -> when (span.style) {
                Typeface.BOLD -> "<b>"
                Typeface.ITALIC -> "<i>"
                Typeface.BOLD_ITALIC -> "<b><i>"
                else -> null
            }

            is RubySpan -> when (span.position) {
                TextAnnotation.POSITION_BEFORE -> "<ruby style='ruby-position:over;'>"
                TextAnnotation.POSITION_AFTER -> "<ruby style='ruby-position:under;'>"
                TextAnnotation.POSITION_UNKNOWN -> "<ruby style='ruby-position:unset;'>"
                else -> null
            }

            is UnderlineSpan -> "<u>"
            is TextEmphasisSpan -> {
                val style = getTextEmphasisStyle(span.markShape, span.markFill)
                val position = getTextEmphasisPosition(span.position)
                Util.formatInvariant(
                    "<span style='-webkit-text-emphasis-style:%1\$s;text-emphasis-style:%1\$s;-webkit-text-emphasis-position:%2\$s;text-emphasis-position:%2\$s;display:inline-block;'>",
                    style, position
                )
            }

            else -> null
        }
    }

    @OptIn(UnstableApi::class)
    private fun getClosingTag(span: Any): String? {
        return when (span) {
            is StrikethroughSpan,
            is ForegroundColorSpan,
            is BackgroundColorSpan,
            is HorizontalTextInVerticalContextSpan,
            is AbsoluteSizeSpan,
            is RelativeSizeSpan,
            is TextEmphasisSpan -> "</span>"

            is TypefaceSpan -> if (span.family != null) "</span>" else null
            is StyleSpan -> when (span.style) {
                Typeface.BOLD -> "</b>"
                Typeface.ITALIC -> "</i>"
                Typeface.BOLD_ITALIC -> "</i></b>"
                else -> null
            }

            is RubySpan -> "<rt>${escapeHtml(span.rubyText)}</rt></ruby>"
            is UnderlineSpan -> "</u>"
            else -> null
        }
    }

    @OptIn(UnstableApi::class)
    private fun getTextEmphasisStyle(shape: Int, fill: Int): String {
        val builder = StringBuilder()
        when (fill) {
            TextEmphasisSpan.MARK_FILL_FILLED -> builder.append("filled ")
            TextEmphasisSpan.MARK_FILL_OPEN -> builder.append("open ")
        }
        when (shape) {
            TextEmphasisSpan.MARK_SHAPE_CIRCLE -> builder.append("circle")
            TextEmphasisSpan.MARK_SHAPE_DOT -> builder.append("dot")
            TextEmphasisSpan.MARK_SHAPE_SESAME -> builder.append("sesame")
            TextEmphasisSpan.MARK_SHAPE_NONE -> builder.append("none")
            else -> builder.append("unset")
        }
        return builder.toString()
    }

    @OptIn(UnstableApi::class)
    private fun getTextEmphasisPosition(position: Int): String {
        return if (position == TextAnnotation.POSITION_AFTER) "under left" else "over right"
    }

    private fun SparseArray<Transition>.getOrCreate(key: Int): Transition {
        var transition = this.get(key)
        if (transition == null) {
            transition = Transition()
            this.put(key, transition)
        }
        return transition
    }

    private fun escapeHtml(text: CharSequence) =
        NEWLINE_REGEX.replace(Html.escapeHtml(text), "<br>")

    data class HtmlAndCss(
        val html: String,
        val cssRuleSets: Map<String, String>
    )

    private data class SpanInfo(
        val start: Int,
        val end: Int,
        val openingTag: String,
        val closingTag: String
    ) {
        companion object {
            val FOR_OPENING_TAGS = compareByDescending<SpanInfo> { it.end }
                .thenBy { it.openingTag }
                .thenBy { it.closingTag }

            val FOR_CLOSING_TAGS = compareByDescending<SpanInfo> { it.start }
                .thenByDescending { it.openingTag }
                .thenByDescending { it.closingTag }
        }
    }

    private class Transition {
        val spansAdded = mutableListOf<SpanInfo>()
        val spansRemoved = mutableListOf<SpanInfo>()
    }
}